src/Raven.Server/Documents/Patch/ScriptRunner.cs

```diff
@@ -1168,10 +1171,27 @@ public JsValue PutDocument(JsValue self, JsValue[] args)
                 }
             }
 
-            private string GenerateIdForShard(string id)
+            private string GenerateIdForShard(ShardedDocumentDatabase shardedDocumentDatabase, string id)
             {
-                if (id?[^1] != _database.IdentityPartsSeparator)
+                if (string.IsNullOrWhiteSpace(id))
+                {
+                    id = $"{Guid.NewGuid()}${OriginalDocumentId}";
+                }
+                
+                if (id[^1] != _database.IdentityPartsSeparator)
+                {
+                    var config = shardedDocumentDatabase.ShardingConfiguration;
+                    var originalBucket = ShardHelper.GetBucketFor(config, _docsCtx.Allocator, OriginalDocumentId);
+                    var currentBucketId = ShardHelper.GetBucketFor(config, _docsCtx.Allocator, id);
+                    if (originalBucket != currentBucketId)
+                    {
+                        throw new ShardedPatchBehaviorViolationException(
+                            $"The original ID '{OriginalDocumentId}' isn't in the same bucket as the requested ID '{id}'.{Environment.NewLine}" +
+                            $"To Ensure they will be in the same bucket use the '$' convention.{Environment.NewLine}" +
```

Comment:
To ensure operation will be performed on the same bucked, use anchored ID with '$'